tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala

```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
   /** Removes elem from the scope, as the scope only contains types of free variables */
   def markAsBound(elem: TlaEx): TBuilderInternalState[Unit] = State[TBuilderContext, Unit] { mi: TBuilderContext =>
     require(elem.isInstanceOf[NameEx])
-    (mi.copy(nameScope = mi.nameScope - elem.asInstanceOf[NameEx].name), ())
+    (mi.copy(freeNameScope = mi.freeNameScope - elem.asInstanceOf[NameEx].name), ())
   }
 
+  /** Performs shadowing checks for ternary operators which introduce bound variables */
+  def boundVarIntroductionTernary(
+      rawMethod: (TlaEx, TlaEx, TlaEx) => TlaEx // order: variable, set, expression
+    )(variable: TBuilderInstruction,
+      set: TBuilderInstruction,
+      expr: TBuilderInstruction): TBuilderInstruction = for {
+    setEx <- set
+    usedInSet <- allUsed // x may not appear as bound or free in set
+    exprEx <- expr
+    boundAfterExpr <- allBound // x may not appear as bound in expr
+    varEx <- variable
+    _ <- markAsBound(varEx)
+    // variable is shadowed iff boundAfterVar \subseteq usedInSet \union boundAfrerExpr
+    boundAfterVar <- allBound
+    diff = (boundAfterVar -- usedInSet) -- boundAfterExpr
```

Comment:
Given the comment, why not check for subset relation and then test that boolearather than computing the difference and then checking whether that is empty?

Reply:
I mean, they're equivalent. 
```
A \subseteq B <=> A \ B \subseteq B \ B <=> A \ B \subseteq {} <=> A \ B = {}
```

Reply:
I've changed the tests to be the same as the comments though, might as well.